### PR TITLE
Access control: Using RBAC to filter users in list view that you have read access to

### PR DESCRIPTION
**Pull request by eleijonmarck**
**What this PR does / why we need it**:
Currently all users are show in the listing in server admin. This PR filters users listed based on read access

- [x] test with user with only read permission for all and read permission on a specific user
- [x] add tests inplace for listing a specfic set of users, so we know the functionality works.

Searchusers is used in:
`ldapsync.go`

**Which issue(s) this PR fixes**:
https://github.com/grafana/grafana-enterprise/issues/2588

Fixes #

**Special notes for your reviewer**:

### CODE REVIEW

File: pkg/services/sqlstore/user.go [hunk 1]
```diff
@@ -620,6 +622,21 @@ func (ss *SQLStore) SearchUsers(ctx context.Context, query *models.SearchUsersQu
 			whereParams = append(whereParams, query.OrgId)
 		}
 
+		// user only sees the users in which it has user read access too
+		if ss.Cfg.IsFeatureToggleEnabled(featuremgmt.FlagAccesscontrol) && query.SignedInUser != nil {
+			acFilter, err := accesscontrol.Filter(
+				query.SignedInUser,
+				"u.id",
+				"users:id:",
+				accesscontrol.ActionOrgUsersRead,
```
Comment (eleijonmarck) [inline pkg/services/sqlstore/user.go:631]: I wanted to use `accescontrol.ActionUsersRead` however when using that permission, no users are shown.

File: pkg/services/sqlstore/user.go [hunk 2]
```diff
@@ -620,6 +621,21 @@ func (ss *SQLStore) SearchUsers(ctx context.Context, query *models.SearchUsersQu
 			whereParams = append(whereParams, query.OrgId)
 		}
 
+		// user only sees the users in which it has user read access too
+		if ss.Cfg.IsFeatureToggleEnabled(featuremgmt.FlagAccesscontrol) && query.SignedInUser != nil {
+			acFilter, err := ac.Filter(
+				query.SignedInUser,
+				"u.id",
+				"users:id:",
+				ac.ActionOrgUsersRead,
```
Comment (gamab) [inline pkg/services/sqlstore/user.go:630]: It seems that the search user endpoint is using this and is protected with `ac.EvalPermission(ac.ActionUsersRead, ac.ScopeGlobalUsersAll))`. Did you try the following?
```suggestion
				"global.users:id:",
				ac.ActionUsersRead,
```

**Review by gamab — COMMENTED** [review]
`SearchUsers` seems to be used in `ldapsync.go` and `comments>handlers.go` can you double check you are not breaking something there? 
Maybe check every place `SearchUsersQuery` is used and double check the `SignedInUser` is correctly set. In some cases, for functions that need no restriction, we had to create a `tmpUser` granting him the required permissions. (Note that this is not the best and we will have to figure out a correct pattern for this specific case)

File: pkg/services/sqlstore/user.go (hunk 2 above)
Comment (eleijonmarck) [reply pkg/services/sqlstore/user.go:630]: ahhh i did not do it with `global` 😶‍🌫️ . then it probably works

**Comment by eleijonmarck** [comment]
> `SearchUsers` seems to be used in ldapsync.go and comments>handlers.go can you double check you are not breaking something there?

also,  i really need to get to the bottom of how to setup enterprise properly in vscode

**Comment by eleijonmarck** [comment]
> `SearchUsers` seems to be used in `ldapsync.go` and `comments>handlers.go` can you double check you are not breaking something there? Maybe check every place `SearchUsersQuery` is used and double check the `SignedInUser` is correctly set. In some cases, for functions that need no restriction, we had to create a `tmpUser` granting him the required permissions. (Note that this is not the best and we will have to figure out a correct pattern for this specific case)

Btw; if we leave it as a dangling pointer if can be always amitted from the query. So it is not needed in the queries.

**Comment by eleijonmarck** [comment]
remeber to merge - https://github.com/grafana/grafana-enterprise/pull/3244

File: pkg/services/sqlstore/user.go [hunk 3]
```diff
@@ -620,6 +621,21 @@ func (ss *SQLStore) SearchUsers(ctx context.Context, query *models.SearchUsersQu
 			whereParams = append(whereParams, query.OrgId)
 		}
 
+		// user only sees the users in which it has user read access too
```
Comment (IevaVasiljeva) [inline pkg/services/sqlstore/user.go:624]:
```suggestion
		// user only sees the users for which it has read permissions
```

File: pkg/services/sqlstore/user.go [hunk 4]
```diff
@@ -620,6 +621,21 @@ func (ss *SQLStore) SearchUsers(ctx context.Context, query *models.SearchUsersQu
...
+		if ss.Cfg.IsFeatureToggleEnabled(featuremgmt.FlagAccesscontrol) && query.SignedInUser != nil {
```
Comment (IevaVasiljeva) [inline pkg/services/sqlstore/user.go:625]: question: should we return an error or empty result if `SignedInUser` isn't provided? It's quite easy to forget to set this field, which could then result in us listing more users than the caller has access to.

**Review by IevaVasiljeva — COMMENTED** [review]
Looks good, but I think you now need to remove the `ScopeGlobalUsersAll` scope from [user listing endpoint](https://github.com/grafana/grafana/blob/main/pkg/api/api.go#L178) to allow people with partial user listing permissions to use this endpoint.

It'd also be good to add some tests for user listing when the caller is only allowed to list a subset of the existing users.

**Comment by eleijonmarck** [comment]
> Looks good, but I think you now need to remove the `ScopeGlobalUsersAll` scope from [user listing endpoint](https://github.com/grafana/grafana/blob/main/pkg/api/api.go#L178) to allow people with partial user listing permissions to use this endpoint.

You mean remove the scope entierly since now we serve it as not seeing all users. So no really need to have that scope at all. 


> It'd also be good to add some tests for user listing when the caller is only allowed to list a subset of the existing users.

I here you on the `tests` cases that should be inplace. Let me add that to tasks todo.

File: pkg/services/sqlstore/user.go (hunk 4 above)
Comment (eleijonmarck) [reply pkg/services/sqlstore/user.go:625]: then we can check that it needs to be set.

**Comment by gamab** [comment]
Just went over this one again, I think my previous comment is still true:
> SearchUsers seems to be used in ldapsync.go and comments>handlers.go can you double check you are not breaking something there?
Maybe check every place SearchUsersQuery is used and double check the SignedInUser is correctly set. In some cases, for functions that need no restriction, we had to create a tmpUser granting him the required permissions. (Note that this is not the best and we will have to figure out a correct pattern for this specific case)

At least `ldapsync` would fail due to the lack of signed in user:
> logger=ldap.sync t=2022-05-12T17:03:30.020593586+02:00 level=error msg="LDAP Sync on startup failed" error="missing permissions"

> tests pass for ldapsync
This is a bit misguiding, they pass against a mock SQLStore.

**Comment by gamab** [comment]
Still missing the signed in user there: https://github.com/grafana/grafana/blob/1c679e814b1e031a351c3d2c066eb85d8633150d/pkg/services/comments/handlers.go#L149

> curl --basic -u "admin:*****" -X POST -d '{"objectType": "annotation", "objectID": "3"}' -H 'Content-Type: application/json' http://localhost:3000/api/comments/get
{"error":"missing permissions","message":"internal error","traceID":"00000000000000000000000000000000"}

**Review by gamab — APPROVED** [review]

**Comment by eleijonmarck** [comment]
great work

**Pull request merged**